src/funnels/Swapper.sol

```diff
@@ -0,0 +1,243 @@
+    event Quit  (address indexed usr, uint256 wad);
+
+    struct Weight {
+        address box;
+        uint96 wad; // percentage in WAD such that 1 WAD = 100%
+    }
+
+    constructor(address _nst, address _gem, address _uniV3Router) {
+        nst = _nst;
+        gem = _gem;
+        uniV3Router = _uniV3Router;
+
+        GemLike(nst).approve(uniV3Router, type(uint256).max);
+        GemLike(gem).approve(uniV3Router, type(uint256).max);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "Swapper/not-authorized");
+        _;
+    }
+
+    // permissionned to whitelisted facilitators
+    modifier toll { 
+        require(buds[msg.sender] == 1, "Swapper/non-facilitator"); 
+        _;
+    }
+
+    // permissionned to whitelisted keepers
+    modifier keeper { 
+        require(keepers[msg.sender] == 1, "Swapper/non-keeper"); 
+        _;
+    }
+
+    uint256 internal constant WAD = 10 ** 18;
+
+    address public immutable uniV3Router;
+    address public immutable nst;
+    address public immutable gem;
+
+    function rely(address usr)   external auth { wards[usr]   = 1; emit Rely(usr); }
+    function deny(address usr)   external auth { wards[usr]   = 0; emit Deny(usr); }
+    function kiss(address usr)   external auth { buds[usr]    = 1; emit Kissed(usr); }
+    function diss(address usr)   external auth { buds[usr]    = 0; emit Dissed(usr); }
+    function permit(address usr) external toll { keepers[usr] = 1; emit Permit(usr); }
+    function forbid(address usr) external toll { keepers[usr] = 0; emit Forbid(usr); }
+
+    function file(bytes32 what, uint256 data) external auth {
+        if      (what == "maxIn")   maxIn  = data;
+        else if (what == "maxOut")  maxOut = data;
+        else if (what == "wantIn")  wantIn = data;
+        else if (what == "wantOut") wantOut = data;
+        else if (what == "hop")     hop    = data;
+        else if (what == "fee")     fee    = data;
+        else revert("Swapper/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        if (what == "buffer") buffer = data;
+        else revert("Swapper/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, address data, uint256 val) external auth {
+        if (what == "box") boxes[data] = val;
+        else revert("Swapper/file-unrecognized-param");
+        emit File(what, data, val);
+    }
+
+    function setWeights(Weight[] memory newWeights) external toll {
+        uint256 cumPct;
+        uint256[] memory arr = new uint256[](newWeights.length);
+        for(uint256 i; i < newWeights.length;) {
+            (address _box, uint256 _pct) = (newWeights[i].box, uint256(newWeights[i].wad));
+            require(boxes[_box] == 1, "Swapper/invalid-box");
```

Comment:
Not sure we need this if we have the check at `getWeightAt`.
It can actually be convenient in some cases for the facilitator to `setWeights` before the executive actually sets the boxes.